saleor/graphql/product/mutations/products.py

```diff
@@ -359,12 +360,23 @@ def perform_mutation(cls, _root, info, collection_id, products):
         collection = cls.get_node_or_error(
             info, collection_id, field="collection_id", only_type=Collection
         )
+        products_without_variants = get_products_without_variants(products)
+        if products_without_variants:
+            raise ValidationError(
+                {
+                    "products": ValidationError(
+                        f"Cannot manage products without variants. "
+                        f"Products ids: {products_without_variants}.",
+                        code=ProductErrorCode.CANNOT_MANAGE_PRODUCT_WITHOUT_VARIANT,
+                    )
+                }
+            )
```

Comment:
Maybe something like this:

```
products = cls.get_nodes_or_error(products, "products", Product)
clean_products(products)
# rest of mutation logic
```

And above we'd add a new function in this mutation:
```
def clean_products(...):
    products_without_variants = get_products_without_variants(products)
    if products_without_variants:
        raise ValidationError(...)
```

That would be consistent with the rest of the codebase.